File: rust/tests/power/ina226_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::power::{Ina226Full, BOL};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond {
            println!("PASS {}", $label);
            $passed += 1;
        } else {
            println!("FAIL {}", $label);
            $failed += 1;
        }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x40);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut chip = Ina226Full::new(dev, addr, 0.1, 2.0).expect("init INA226");

    let mut passed = 0i32;
    let mut failed = 0i32;

    check_true!(
        chip.manufacturer_id().unwrap() == 0x5449,
        "manufacturer_id",
        passed,
        failed
    );
    check_true!(
        chip.die_id().unwrap() == 0x2260,
        "die_id",
        passed,
        failed
    );

    chip.configure(0, 4, 4, 7).unwrap();

    let v = chip.voltage().unwrap();
    check_true!(v >= 0.0 && v <= 40.0, "voltage_range", passed, failed);

    let sv = chip.shunt_voltage().unwrap();
    check_true!(sv >= -0.082 && sv <= 0.082, "shunt_voltage_range", passed, failed);

    let i = chip.current().unwrap();
    check_true!(i >= -2.0 && i <= 2.0, "current_range", passed, failed);

    let p = chip.power().unwrap();
    check_true!(p >= 0.0 && p <= 80.0, "power_range", passed, failed);

    check_true!(
        chip.conversion_ready().is_ok(),
        "conversion_ready_ok",
        passed,
        failed
    );
    check_true!(
        chip.overflow().is_ok(),
        "overflow_ok",
        passed,
        failed
    );

    chip.set_alert(BOL, 5.0, false, false).unwrap();
    let flags = chip.alert_flags().unwrap();
    check_true!(flags & BOL != 0, "alert_flags_bol_set", passed, failed);

    chip.shutdown().unwrap();
    chip.wake().unwrap();
    check_true!(true, "shutdown_wake", passed, failed);

    chip.reset().unwrap();
    check_true!(true, "reset", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/power/ina226_test/main.go
//go:build linux && !tinygo

// INA226 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the INA226 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/power"
	"github.com/tuhde/Periph/go/periph/transport"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x40"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	tr, err := transport.NewI2CTransport(bus, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport:", err)
		os.Exit(2)
	}
	defer tr.Close()

	chip, err := power.NewINA226Full(tr, 0.1, 2.0)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	mfr, err := chip.ManufacturerID()
	check("manufacturer_id", err == nil && mfr == 0x5449)

	die, err := chip.DieID()
	check("die_id", err == nil && die == 0x2260)

	if err := chip.Configure(0, 4, 4, 7); err != nil {
		fmt.Fprintln(os.Stderr, "configure:", err)
		os.Exit(2)
	}

	v, err := chip.Voltage()
	check("voltage_range", err == nil && v >= 0.0 && v <= 40.0)

	sv, err := chip.ShuntVoltage()
	check("shunt_voltage_range", err == nil && sv >= -0.082 && sv <= 0.082)

	c, err := chip.Current()
	check("current_range", err == nil && c >= -2.0 && c <= 2.0)

	p, err := chip.Power()
	check("power_range", err == nil && p >= 0.0 && p <= 80.0)

	_, err = chip.ConversionReady()
	check("conversion_ready_ok", err == nil)

	_, err = chip.Overflow()
	check("overflow_ok", err == nil)

	if err := chip.SetAlert(power.BOL, 5.0, false, false); err != nil {
		fmt.Fprintln(os.Stderr, "set_alert:", err)
		os.Exit(2)
	}
	flags, err := chip.AlertFlags()
	check("alert_flags_bol_set", err == nil && flags&power.BOL != 0)

	if err := chip.Shutdown(); err != nil {
		fmt.Fprintln(os.Stderr, "shutdown:", err)
		os.Exit(2)
	}
	if err := chip.Wake(); err != nil {
		fmt.Fprintln(os.Stderr, "wake:", err)
		os.Exit(2)
	}
	check("shutdown_wake", true)

	if err := chip.Reset(); err != nil {
		fmt.Fprintln(os.Stderr, "reset:", err)
		os.Exit(2)
	}
	check("reset", true)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
